Add notification scheduling logic

Garuda-1 opened the pull request:
Added a scheduled task to fetch all notifications in range from negative infinity
to current time plus 60 seconds. Each fetched notification causes subtasks to be
scheduled to write a specific user about todo.

Garuda-1 committed:
Add notification scheduling logic

Added a scheduled task to fetch all notifications in range from negative infinity
to current time plus 60 seconds. Each fetched notification causes subtasks to be
scheduled to write a specific user about todo.

Deletrius366 reviewed (commented):
Can you provide some information about testing done? Anyway, our global task includes writing tests for our logic parts. From now, code is complex enough to verify only by "eyes". My idea is to split this quite big commit (PR) into smaller one. Anyway, I would like to see for the first *Repository testing (queries testing), for the second - TaskFetchingService implementation WITH unit testing for it. I'm open for discussion about details.

Garuda-1 committed:
Add unit tests with JPA for TaskFetchingService

Added test scenarios to check that appropriate notifications are fetched from the
database and cause actual message to be sent.

Garuda-1 committed:
Add missing trailing newlines

Deletrius366 commented on src/test/java/ru/itmo/sd/deadliner2bot/service/TaskFetchingServiceJpaTest.java:
```diff
@@ -0,0 +1,178 @@
+        return chatRepository.save(chat);
+    }
+
+    private DailyNotification createDailyNotification(Chat chat, LocalDateTime notificationTime) {
```
Looks like almost the same (with slightly different signature) create* functions will be used during bot work (not only for testing). So I suggest to move them to corresponding services in source part.

Garuda-1 replied:
I am highly unsure about this. I suggest refactoring as soon as required.

Deletrius366 commented on src/main/java/ru/itmo/sd/deadliner2bot/dto/TodoDto.java:
```diff
@@ -0,0 +1,18 @@
+
+@Data
+@Builder
+public class TodoDto {
```
Why can't we use Todo class? The same for TodoNotification

Garuda-1 replied:
Model classes have a complicated lifetime managed by JPA. We need something persistent instead.

Garuda-1 replied:
Moreover, frequently we don't require fields those reference other entities, so it's better to store data in more lightweight DTOs.

Deletrius366 replied:
> Model classes have a complicated lifetime managed by JPA. We need something persistent instead.

I'm still not understanding, why we need them. Can you clarify please, what are the problems with lifetime? I see, that you don't want to use JPA objects in Runnable tasks, but is it important?

> Moreover, frequently we don't require fields those reference other entities, so it's better to store data in more lightweight DTOs.

2 instances worse than 1, so if we can escape additional objects, we should do it.

Garuda-1 replied:
There is no point to store in long-term instances (like `Runnables` for `SchedulingExecutorService`) data with unnecessary fields. 
It is a common practice to use DTOs alongside entities:
https://www.baeldung.com/java-dto-pattern
https://www.baeldung.com/entity-to-and-from-dto-for-a-java-spring-application

Deletrius366 replied:
I've read. As for me, this is not our case. DTO's are used to transfer data to other serivices - in such case amount of data is important. We are not transfering, no serializing or JSON messages, only using by runnables. I recommend to avoid these classes for now. Also, do you use TodoNotificationDto?

Deletrius366 commented on src/test/java/ru/itmo/sd/deadliner2bot/service/TaskFetchingServiceJpaTest.java:
```diff
@@ -0,0 +1,178 @@
+
+    @Test
+    @DisplayName(value = "Daily notification added in scheduled thread pool")
+    public void dailyNotificationAddedInScheduledThreadPool() {
```
Will be good to check, that notifications were removed after fetching.

Garuda-1 replied:
Good point.

Garuda-1 replied:
Added in the follow-up.

Deletrius366 commented on src/main/java/ru/itmo/sd/deadliner2bot/messages/MessageFormatter.java:
```diff
@@ -0,0 +1,47 @@
+        StringJoiner lines = new StringJoiner("\n");
+        lines.add(messageTemplatesProperties.getProperty("daily-notifications-header"));
+        todoDtos.stream()
+                .sorted(Comparator.nullsLast(Comparator.comparing(TodoDto::getEndTime)))
```
If I am right, here list has been already sorted (we got sorted TreeSet from database, after this performed operations in TreeSet order). Anyway, let's keep invariant, that notifications are sorted everywhere include database.

Garuda-1 replied:
This class operates with DTOs and thus all ordering guarantees are missing. Better to sort before rendering, ordering doesn't matter before.

Deletrius366 replied:
From this point, I think we should replace Set in JPA classes by List

Garuda-1 replied:
You should investigate a bit more, using List is actually more expensive than set.

Deletrius366 replied:
Ok

Deletrius366 reviewed (commented):
All in all, LGTM.

Garuda-1 committed:
Add assertions that notification is removed from the database after fetch
